[cpu] implement flash attention backward

Valentine233 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105131
* #107128
* #104863
* __->__ #104693
* #103826
* #104584
* #104583

Feature RFC: https://github.com/pytorch/rfcs/pull/56.

The flash attention CPU kernel is added, for backward path FP32. Parallelization is on the dimensions of batch size and head number.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

Valentine233 committed:
implement flash attention backward

[ghstack-poisoned]

mingfeima commented:
you may cherry pick https://github.com/pytorch/pytorch/pull/104583 and https://github.com/pytorch/pytorch/pull/104584 to this stack and replace `_mkl_gemm` with `cpublas::gemm`, then the mkl platform specific CI fails will pass.

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -370,6 +370,231 @@ void cpu_flash_attention(
+  int64_t num_head = query.size(2);
+  int64_t headSize = query.size(3);
+
+  // Strides
+  int64_t qStrideB = query.stride(0);
+  int64_t qStrideM = query.stride(1);
+  int64_t qStrideH = query.stride(2);
+  int64_t kStrideB = key.stride(0);
+  int64_t kStrideN = key.stride(1);
+  int64_t kStrideH = key.stride(2);
+  int64_t vStrideB = value.stride(0);
+  int64_t vStrideN = value.stride(1);
+  int64_t vStrideH = value.stride(2);
+  int64_t oStrideB = out.stride(0);
+  int64_t oStrideM = out.stride(1);
+  int64_t oStrideH = out.stride(2);
+  int64_t lStrideB = logsumexp.stride(0);
+  int64_t lStrideM = logsumexp.stride(1);
+  int64_t lStrideH = logsumexp.stride(2);
```
Can we simplify the data ptr access with `TensorAccessor`? We don't have to retrieve all these strides vars and also it is simpler to get the data ptr with the index variables. Not sure about the perf impact though.

Valentine233 replied:
Good idea, thanks! As the change is relatively big and the perf impact is also needed to verify, is it OK to do this change in another separate PR?

jgong5 replied:
Sure.

drisspg commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -403,8 +628,51 @@ void flash_attention_kernel_impl(
+          cum_seq_q, cum_seq_k, max_q, max_k, dropout_p,
+          is_causal, philox_seed, philox_offset, scale);
+    });
+    // AT_DISPATCH_CASE(ScalarType::BFloat16, [&] {
```
Do we still need this?

Valentine233 committed:
Update on "[cpu] implement flash attention backward"


Feature RFC: https://github.com/pytorch/rfcs/pull/56.

The flash attention CPU kernel is added, for backward path FP32. Parallelization is on the dimensions of batch size and head number.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement flash attention backward"


Feature RFC: https://github.com/pytorch/rfcs/pull/56.

The flash attention CPU kernel is added, for backward path FP32. Parallelization is on the dimensions of batch size and head number.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 commented:
@pytorchbot merge
